Following are the links to the data used in the notebook: 
[Telecom customer churn](https://www.kaggle.com/blastchar/telco-customer-churn)

[RBI Data](https://www.kaggle.com/yash1158/rbi-g-sec-turnover-data)

Note : Incase the dataset is updated the results of operations carried out on the dataset might differ than in this notebook.

### Installing Pandas

pip install pandas

conda install pandas

In [ ]:
# Importing pandas

import pandas as pd

### Creating a dataframe from scratch

In [ ]:
df = pd.DataFrame({'Name': ['Max','Sam','James'],
                   'Age': [25,24,29],
                   'Weight(in kg)': [75,68,70]})

df

Name  Age  Weight(in kg)
0    Max   25             75
1    Sam   24             68
2  James   29             70

### Reading in Data using pandas

In [ ]:
# Reading in data
data_1 = pd.read_csv("/content/drive/My Drive/Kaggle Datasets /RBI Data.csv")
data_2 = pd.read_csv("/content/drive/My Drive/Kaggle Datasets /WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [ ]:
# First five rows of the data using the head function
data_1.head()

Week Ended  Central Government Dated Securities  ... 364 - Day   RBI
0  23-Aug-2019                              5342.68  ...    150.85     -
1  16-Aug-2019                              2868.51  ...    110.90     -
2  09-Aug-2019                              6772.56  ...    476.98     -
3  02-Aug-2019                              5337.15  ...    194.61     -
4  26-Jul-2019                              5311.22  ...    211.26  0.25

[5 rows x 9 columns]

In [ ]:
# Last five rows of the data using the tail function
data_1.tail()

Week Ended  Central Government Dated Securities  ... 364 - Day   RBI
1116  10-Apr-1998                                52.34  ...     12.07  0.46
1117  03-Apr-1998                                37.42  ...     15.22  1.65
1118          NaN                                  NaN  ...       NaN   NaN
1119          NaN                                  NaN  ...       NaN   NaN
1120          NaN                                  NaN  ...       NaN   NaN

[5 rows x 9 columns]

In [ ]:
# Shape of the dataframe

data_1.shape

(1121, 9)

The dataset has 1117 rows and 9 columns.

### Dropping missing values

In [ ]:
# Handling missing values

data_1.isnull().sum()

Week Ended                             4
Central Government Dated Securities    4
State Government Dated Securities      4
Cash Management Bills                  4
14 - Day                               4
91-Day                                 4
182 - Day                              4
364 - Day                              4
RBI                                    4
dtype: int64

In [ ]:
# Dropping rows with missing values 
data_1.dropna()

Week Ended  Central Government Dated Securities  ... 364 - Day   RBI
0     23-Aug-2019                              5342.68  ...    150.85     -
1     16-Aug-2019                              2868.51  ...    110.90     -
2     09-Aug-2019                              6772.56  ...    476.98     -
3     02-Aug-2019                              5337.15  ...    194.61     -
4     26-Jul-2019                              5311.22  ...    211.26  0.25
...           ...                                  ...  ...       ...   ...
1113  01-May-1998                                53.46  ...      9.92  0.00
1114  24-Apr-1998                                72.22  ...     10.02  0.02
1115  17-Apr-1998                               104.92  ...     11.66  0.01
1116  10-Apr-1998                                52.34  ...     12.07  0.46
1117  03-Apr-1998                                37.42  ...     15.22  1.65

[1117 rows x 9 columns]

The above command yields the dataframe that contains no missing values but this is a temporary dataframe and it does not drop rows with missing values in the actual dataframe. We can check this by using data.isnull().sum() 

In [ ]:
data_1.isnull().sum()

Week Ended                             4
Central Government Dated Securities    4
State Government Dated Securities      4
Cash Management Bills                  4
14 - Day                               4
91-Day                                 4
182 - Day                              4
364 - Day                              4
RBI                                    4
dtype: int64

We still have all those rows with missing values in the actual dataframe. To make changes to the actual dataframe we need to set the parameter inplace to True.


In [ ]:
data_1.dropna(inplace = True)

In [ ]:
data_1.isnull().sum()

Week Ended                             0
Central Government Dated Securities    0
State Government Dated Securities      0
Cash Management Bills                  0
14 - Day                               0
91-Day                                 0
182 - Day                              0
364 - Day                              0
RBI                                    0
dtype: int64

Finally, we have dropped the rows with missing entries

In [ ]:
data_2.head()

customerID  gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
0  7590-VHVEG  Female              0  ...          29.85        29.85     No
1  5575-GNVDE    Male              0  ...          56.95       1889.5     No
2  3668-QPYBK    Male              0  ...          53.85       108.15    Yes
3  7795-CFOCW    Male              0  ...          42.30      1840.75     No
4  9237-HQITU  Female              0  ...          70.70       151.65    Yes

[5 rows x 21 columns]

In [ ]:
data_2.shape

(7043, 21)

In [ ]:
# Dropping an entire column

data_2.drop(columns = ['SeniorCitizen','Partner'], inplace = True)

In [ ]:
data_2.head()

customerID  gender Dependents  ...  MonthlyCharges TotalCharges Churn
0  7590-VHVEG  Female         No  ...           29.85        29.85    No
1  5575-GNVDE    Male         No  ...           56.95       1889.5    No
2  3668-QPYBK    Male         No  ...           53.85       108.15   Yes
3  7795-CFOCW    Male         No  ...           42.30      1840.75    No
4  9237-HQITU  Female         No  ...           70.70       151.65   Yes

[5 rows x 19 columns]

In [ ]:
data_2.shape

(7043, 19)

In [ ]:
# Dropping specific rows

data_2.drop(index = [1,2,3,6], inplace=True)

In [ ]:
data_2.head()

customerID  gender Dependents  ...  MonthlyCharges TotalCharges Churn
0  7590-VHVEG  Female         No  ...           29.85        29.85    No
4  9237-HQITU  Female         No  ...           70.70       151.65   Yes
5  9305-CDSKC  Female         No  ...           99.65        820.5   Yes
7  6713-OKOMC  Female         No  ...           29.75        301.9    No
8  7892-POOKP  Female         No  ...          104.80      3046.05   Yes

[5 rows x 19 columns]

In [ ]:
# Alternative way of dropping rows

data_2.drop([4,5,7], axis = 0, inplace = True)

In [ ]:
data_2.head()

customerID  gender Dependents  ...  MonthlyCharges TotalCharges Churn
0   7590-VHVEG  Female         No  ...           29.85        29.85    No
8   7892-POOKP  Female         No  ...          104.80      3046.05   Yes
9   6388-TABGU    Male        Yes  ...           56.15      3487.95    No
10  9763-GRSKD    Male        Yes  ...           49.95       587.45    No
11  7469-LKBCI    Male         No  ...           18.95        326.8    No

[5 rows x 19 columns]

### Understanding your data : Data exploration using Pandas 

In [ ]:
# Data type of every column in the data
data_1.dtypes

Week Ended                              object
Central Government Dated Securities    float64
State Government Dated Securities       object
Cash Management Bills                   object
14 - Day                                object
91-Day                                 float64
182 - Day                               object
364 - Day                               object
RBI                                     object
dtype: object

All columns except the week ended column should contain a numeric datatype but have dtype object which means that pandas has classified the data in these columns as string. Thus, we need to convert the dtypes of these columns to numeric dtypes.


In [ ]:
cols = ['State Government Dated Securities','Cash Management Bills','14 - Day','182 - Day','364 - Day','RBI']

data_1[cols] = data_1[cols].apply(pd.to_numeric, errors='coerce')

All numeric columns converted to dtype float64 

In [ ]:
data_1.dtypes

Week Ended                              object
Central Government Dated Securities    float64
State Government Dated Securities      float64
Cash Management Bills                  float64
14 - Day                               float64
91-Day                                 float64
182 - Day                              float64
364 - Day                              float64
RBI                                    float64
dtype: object

In [ ]:
# Summarizing your data using info() method

data_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1117 entries, 0 to 1117
Data columns (total 9 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Week Ended                           1117 non-null   object 
 1   Central Government Dated Securities  1117 non-null   float64
 2   State Government Dated Securities    1115 non-null   float64
 3   Cash Management Bills                127 non-null    float64
 4   14 - Day                             161 non-null    float64
 5   91-Day                               1117 non-null   float64
 6   182 - Day                            869 non-null    float64
 7   364 - Day                            1116 non-null   float64
 8   RBI                                  1093 non-null   float64
dtypes: float64(8), object(1)
memory usage: 87.3+ KB


In [ ]:
# Summarizing your data for further analysis using basic statistical metrics (useful for numeric data)

data_1.describe()

Central Government Dated Securities  ...          RBI
count                          1117.000000  ...  1093.000000
mean                           1571.573617  ...    20.355865
std                            1917.006727  ...    51.506649
min                               3.930000  ...     0.000000
25%                             240.640000  ...     0.170000
50%                             738.600000  ...     2.160000
75%                            2459.200000  ...    15.150000
max                           15351.130000  ...   887.150000

[8 rows x 8 columns]

The decsribe method very well summarizes the numeric data using metrics like mean, median, mode and quartiles. But, how do we summarize the data when we have categorical columns and we need further understand the nature of our data for further analysis.

In [ ]:
data_2.head()

customerID  gender Dependents  ...  MonthlyCharges TotalCharges Churn
0   7590-VHVEG  Female         No  ...           29.85        29.85    No
8   7892-POOKP  Female         No  ...          104.80      3046.05   Yes
9   6388-TABGU    Male        Yes  ...           56.15      3487.95    No
10  9763-GRSKD    Male        Yes  ...           49.95       587.45    No
11  7469-LKBCI    Male         No  ...           18.95        326.8    No

[5 rows x 19 columns]

We have many categorical columns in this dataset. Let us observe the unique categories of all these categorical columns.

In [ ]:
# nunique() function tells us about the number of distinct categories for each column

data_2.nunique()

customerID          7036
gender                 2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1583
TotalCharges        6524
Churn                  2
dtype: int64

In [ ]:
# value_counts() function goes one step further and tells us the number of entries for each category for every column.
# But due to the nature of the output of the function this can be applied to a single column at a time as follows

data_2['gender'].value_counts()

Male      3551
Female    3485
Name: gender, dtype: int64

### Slicing and Indexing Data

In [ ]:
# .loc[] is a widely used function for slicing through dataframes

# Suppose we want to slice the dataframe through the gender column 

male = data_2.loc[data_2['gender'] == 'Male']
male.head()

customerID gender Dependents  ...  MonthlyCharges TotalCharges Churn
9   6388-TABGU   Male        Yes  ...           56.15      3487.95    No
10  9763-GRSKD   Male        Yes  ...           49.95       587.45    No
11  7469-LKBCI   Male         No  ...           18.95        326.8    No
12  8091-TTVAX   Male         No  ...          100.35       5681.1    No
13  0280-XJGEX   Male         No  ...          103.70       5036.3   Yes

[5 rows x 19 columns]

In [ ]:
# Further we can also slice a dataframe into another dataframe with only required columns

new_data_2 = data_2.loc[:,['customerID','gender','Dependents']]
new_data_2.head()

customerID  gender Dependents
0   7590-VHVEG  Female         No
8   7892-POOKP  Female         No
9   6388-TABGU    Male        Yes
10  9763-GRSKD    Male        Yes
11  7469-LKBCI    Male         No

The above code would give the specified columns and all the rows. We can slice both over rows and columns simultaneously as well.

In [ ]:
# Slicing the dataframe to a new dataframe with first 100 rows and first 3 columns

new_data_3 = data_2.loc[:100,['customerID','gender','Dependents']]
new_data_3.tail()

customerID  gender Dependents
96   9803-FTJCG    Male        Yes
97   0278-YXOOG    Male         No
98   3212-KXOCR    Male         No
99   4598-XLKNJ  Female         No
100  6380-ARCEH    Male         No

Another function similar to .loc[] is the .iloc[] function ('i' stands for index), since i iterates over index.

In [ ]:
# Slicing the dataframe to a new dataframe with first 100 rows and first 3 columns using .iloc[] function

new_data_3_1 = data_2.iloc[:100,:3]
new_data_3_1.tail()

customerID  gender Dependents
102  7123-WQUHX    Male         No
103  5386-THSLQ  Female         No
104  3192-NQECA    Male         No
105  6180-YBIQI    Male         No
106  6728-DKUCO  Female        Yes

In [ ]:
new_data_3_2 = data_2.iloc[:100,4:6]
new_data_3_2.tail()

PhoneService     MultipleLines
102          Yes               Yes
103           No  No phone service
104          Yes               Yes
105           No  No phone service
106          Yes               Yes

In [ ]:
new_data_3_3 = data_2.iloc[:100,3:5]
new_data_3_3.tail()

tenure PhoneService
102      38          Yes
103      66           No
104      68          Yes
105       5           No
106      72          Yes

 Note : .iloc[] would iterate over the index and so it will result in the first 100 rows of the dataframe unlike .loc[] which would output upto the 100th row of the dataframe.

### The groupby() and filter() function


The filter() function subsets the dataframe according to the mentioned labels.

In [ ]:
# Filtering the dataframe according to the mentioned items.
# The labels mentioned in the items paramater should be the columns and so one doesn't need to
# mention the axis for the same

data_2.filter(items = ['tenure', 'PhoneService'])

tenure PhoneService
0          1           No
8         28          Yes
9         62          Yes
10        13          Yes
11        16          Yes
...      ...          ...
7038      24          Yes
7039      72          Yes
7040      11           No
7041       4          Yes
7042      66          Yes

[7036 rows x 2 columns]

In [ ]:
# Using the like parameter to filter the dataframe
# The axis = 1 states that filtering is performed across columns and all the columns
# with the characters "Charge" in them will constitute the output

data_2.filter(like = 'Charge', axis = 1)

MonthlyCharges TotalCharges
0              29.85        29.85
8             104.80      3046.05
9              56.15      3487.95
10             49.95       587.45
11             18.95        326.8
...              ...          ...
7038           84.80       1990.5
7039          103.20       7362.9
7040           29.60       346.45
7041           74.40        306.6
7042          105.65       6844.5

[7036 rows x 2 columns]

In [ ]:
# We can also filter the dataframe based on index. 
# Set the 'gender' column as the index
data_2.set_index('gender', inplace = True)

# Filter the index using like parameter and axis = 0 states filtering is across rows
# This will output all the rows with index = Female
data_2.filter(like = 'Female', axis = 0)

customerID Dependents  tenure  ... MonthlyCharges TotalCharges Churn
gender                                 ...                                  
Female  7590-VHVEG         No       1  ...          29.85        29.85    No
Female  7892-POOKP         No      28  ...         104.80      3046.05   Yes
Female  3655-SNQYZ        Yes      69  ...         113.25      7895.15    No
Female  8191-XWSZG         No      52  ...          20.65      1022.95    No
Female  4190-MFLUW        Yes      10  ...          55.20       528.35   Yes
...            ...        ...     ...  ...            ...          ...   ...
Female  0639-TSIQW         No      67  ...         102.95      6886.25   Yes
Female  7750-EYXWZ         No      12  ...          60.65        743.3    No
Female  2569-WGERO         No      72  ...          21.15       1419.4    No
Female  2234-XADUH        Yes      72  ...         103.20       7362.9    No
Female  4801-JZAZL        Yes      11  ...          29.60       346.45    No

[3485 rows x 18 columns]

In [ ]:
data_2.reset_index(inplace = True)

In [ ]:
# Grouping the data by gender and calculating the mean monthly charges gender wise

data_2.groupby(['gender'])['MonthlyCharges'].mean()

gender
Female    65.202956
Male      64.331738
Name: MonthlyCharges, dtype: float64

In [ ]:
# Multiple Grouping
# Grouping the data first by Gender and then by Dependents and calculates average Monthly Charges
# paid by all the categories
data_2.groupby(['gender','Dependents'])['MonthlyCharges'].mean()

gender  Dependents
Female  No            67.624074
        Yes           59.416294
Male    No            66.404575
        Yes           59.595467
Name: MonthlyCharges, dtype: float64

In [ ]:
# Setting up a custom index
# We can set up a column as an index

data_2.set_index(['customerID'], inplace=True)

In [ ]:
data_2.head()

gender Dependents  tenure  ... MonthlyCharges TotalCharges Churn
customerID                             ...                                  
7590-VHVEG  Female         No       1  ...          29.85        29.85    No
7892-POOKP  Female         No      28  ...         104.80      3046.05   Yes
6388-TABGU    Male        Yes      62  ...          56.15      3487.95    No
9763-GRSKD    Male        Yes      13  ...          49.95       587.45    No
7469-LKBCI    Male         No      16  ...          18.95        326.8    No

[5 rows x 18 columns]

In [ ]:
# we can reset the index back to the default state 

data_2.reset_index(inplace = True)

### Sorting

In [ ]:
# Sorting the dataframe by data in a single column
# setting ascending parameter to True will result in sorting the values in ascending order
data_2.sort_values('tenure', ascending= True)

customerID  gender Dependents  ...  MonthlyCharges TotalCharges Churn
481   4472-LVYGI  Female        Yes  ...           52.55                 No
4373  2520-SGTTA  Female        Yes  ...           20.00                 No
6663  4075-WKNIU  Female        Yes  ...           73.35                 No
1333  1371-DWPAZ  Female        Yes  ...           56.05                 No
3324  7644-OMVMY    Male        Yes  ...           19.85                 No
...          ...     ...        ...  ...             ...          ...   ...
6379  6656-GULJQ    Male        Yes  ...           83.55       6093.3    No
779   4992-LTJNE    Male        Yes  ...           55.30       3983.6    No
4208  9031-ZVQPT    Male         No  ...          108.20       7840.6    No
6367  1293-BSEUN  Female        Yes  ...           20.70       1482.3    No
398   7606-BPHHN    Male         No  ...           19.80      1468.75    No

[7036 rows x 19 columns]

In [ ]:
# Sorting can also be performed on calculated results

data_2['InternetService'].value_counts().sort_values(ascending = False)

Fiber optic    3093
DSL            2417
No             1526
Name: InternetService, dtype: int64

### Merging two dataframes

In [ ]:
# Joining the two dataframes and performing the join on the index, both the dataframes have a common index

new_data_merge = pd.merge(new_data_3_1, new_data_3_2, left_index=True, right_index=True)
new_data_merge

customerID  gender Dependents PhoneService     MultipleLines
0    7590-VHVEG  Female         No           No  No phone service
8    7892-POOKP  Female         No          Yes               Yes
9    6388-TABGU    Male        Yes          Yes                No
10   9763-GRSKD    Male        Yes          Yes                No
11   7469-LKBCI    Male         No          Yes                No
..          ...     ...        ...          ...               ...
102  7123-WQUHX    Male         No          Yes               Yes
103  5386-THSLQ  Female         No           No  No phone service
104  3192-NQECA    Male         No          Yes               Yes
105  6180-YBIQI    Male         No           No  No phone service
106  6728-DKUCO  Female        Yes          Yes               Yes

[100 rows x 5 columns]

In [ ]:
# The concat method performs the same function as the merge function above 
# Typical SQL type outer join
# axis parameter set to 1 indicates both dataframes are merged/joined n columns
new_data_concat = pd.concat(objs = [new_data_3_1,new_data_3_2], axis = 1)
new_data_concat

customerID  gender Dependents PhoneService     MultipleLines
0    7590-VHVEG  Female         No           No  No phone service
8    7892-POOKP  Female         No          Yes               Yes
9    6388-TABGU    Male        Yes          Yes                No
10   9763-GRSKD    Male        Yes          Yes                No
11   7469-LKBCI    Male         No          Yes                No
..          ...     ...        ...          ...               ...
102  7123-WQUHX    Male         No          Yes               Yes
103  5386-THSLQ  Female         No           No  No phone service
104  3192-NQECA    Male         No          Yes               Yes
105  6180-YBIQI    Male         No           No  No phone service
106  6728-DKUCO  Female        Yes          Yes               Yes

[100 rows x 5 columns]

In [ ]:
# Suppose, we want to merge to dataframes on a common column among both the dataframes
# Merging dataframes on the 'PhoneService' column and performing a SQL type left join
new_data_3_left = pd.merge(new_data_3_3, new_data_3_2, on='PhoneService', how='left')
new_data_3_left

tenure PhoneService     MultipleLines
0          1           No  No phone service
1          1           No  No phone service
2          1           No  No phone service
3          1           No  No phone service
4          1           No  No phone service
...      ...          ...               ...
8523      72          Yes                No
8524      72          Yes                No
8525      72          Yes               Yes
8526      72          Yes               Yes
8527      72          Yes               Yes

[8528 rows x 3 columns]

In [ ]:
# Using concat function to perform a SQL type inner join i.e. intersection or
# common columns among both the dataframes

new_data_3_inner = pd.concat(objs = [new_data_3_3, new_data_3_2], join='inner')
new_data_3_inner

PhoneService
0             No
8            Yes
9            Yes
10           Yes
11           Yes
..           ...
102          Yes
103           No
104          Yes
105           No
106          Yes

[200 rows x 1 columns]

This tutorial is intended to be a public resource. As such, if you see any glaring inaccuracies or if a critical topic is missing, please feel free to point it out or (preferably) submit a pull request to improve the tutorial. Also, we are always looking to improve the scope of this article. For anything feel free to mail us @ colearninglounge@gmail.com

Author of the article is Vaibhav Desai. You can follow him on [GitHub](https://github.com/desaivaibhav95), [LinkedIn](https://www.linkedin.com/in/vaibhav-desai-0512/), [WordPress](https://dadrdema.wordpress.com/)